tests/callbacks_test.cpp

```diff
@@ -152,6 +152,32 @@ void CallbacksFullFunctionTest(OptimizerType& optimizer,
   REQUIRE(cb.calledStepTaken == calledStepTaken);
 }
 
+template<typename OptimizerType>
+void EarlyStopCallbacksFunctionTest(OptimizerType& optimizer)
+{
+  arma::mat data, testData, shuffledData;
+  arma::Row<size_t> responses, testResponses, shuffledResponses;
+
+  LogisticRegressionTestData(data, testData, shuffledData,
+      responses, testResponses, shuffledResponses);
+  
+  LogisticRegression<> lr(shuffledData, shuffledResponses, 0.5);
+  arma::mat coordinates = lr.GetInitialPoint();
+  LogisticRegressionValidationFunction lrValidation(lr, coordinates);
+
+  EarlyStopAtMinLoss<arma::mat> cb(
+    [&](const arma::mat& /* coordinates */){
+   return lrValidation.Evaluate(testData, testResponses);
+  });
+
+  optimizer.Optimize(lr, coordinates, cb);
+}
+
+TEST_CASE("EarlyStopAtMinLossCallbackTest", "[CallbacksTest]")
+{
+  SMORMS3 smorms3;
+  EarlyStopCallbacksFunctionTest(smorms3);
+}
```

Comment:
It would probably be good if we could also add a test that makes sure that the given lambda will terminate the optimization.  I would suggest this... use, e.g., the Rosenbrock function with a custom lambda callback that terminates whenever any of the coordinates has an absolute value of less than 10.  Then once the optimization terminates, make sure that we are nowhere near the minimum, and that at least one of the coordinates has an absolute value of less than 10.  Let me know if I can clarify that idea---it's not the only possible idea, of course.

Reply:
if you have a code sample of the idea, it would be greate, I am truly not expert in tests :+1:.

Reply:
Ok, here is a rough outline.  It is not perfect.  In fact it may not even compile.  But hopefully it will be enough to get the idea across.

```
// Maybe the name could be shorter...
TEST_CASE("EarlyStopAtMinLossWithCustomLambdaTest", "[CallbacksTest]")
{
  // Use the 50-dimensional Rosenbrock function.
  GeneralizedRosenbrockFunction f(50);
  // Start at some really large point.
  arma::mat coordinates = f.GetInitialPoint();
  f.fill(100.0);

  EarlyStopAtMinLoss<arma::mat> cb(
      [&](const arma::mat& coordinates)
      {
        // Terminate if any coordinate has a value less than 10.
        return arma::any(arma::abs(coordinates) < 10.0);
      });

  SMORMS3 smorms3;
  smorm3.Optimize(f, coordinates, cb);

  // Make sure that we did not get to the optimum.
  for (size_t i = 0; i < coordinates.n_elem; ++i)
    REQUIRE(std::abs(coordinates[i]) >= 3.0);
}
```
Anyway, I think that will work. :+1: